rvm.c

```diff
@@ -141,21 +150,77 @@ rvm_txid_t rvm_txn_begin(rvm_cfg_t cfg)
     return 1;
 }
 
-bool rvm_txn_commit(rvm_cfg_t cfg, rvm_txid_t txid)
+bool check_txn_commit(rvm_cfg_t* cfg, rvm_txid_t txid)
+{
+    LOG(8, ("Checking txn commit\n"));
+
+    // check block table
+    char* block = (char*)malloc(sizeof(char) * cfg->blk_sz);
+    RETURN_ERROR(block == 0, false,
+            ("Failure: Error allocating table\n"));
+
+    struct ibv_mr* block_mr = rmem_create_mr(block, sizeof(char) * cfg->blk_sz);
+
+    LOG(8, ("Getting block table\n"));
+    int err;
+    err = rmem_get(&cfg->rmem, block,
+                block_mr, BLOCK_TBL_ID, cfg->blk_sz);
+    LOG(8, ("Block table fetched\n"));
+
+    RETURN_ERROR(err != 0, false,
+            ("Failure: Error doing RDMA read of block table\n"));
+
+    err = memcmp(block, cfg->blk_tbl, sizeof(char) * cfg->blk_sz);
+
+    RETURN_ERROR(err != 0, false,
+            ("Block table does not match replica\n"));
+
+    //check data blocks
+    int bx;
+    for(bx = 0; bx < BLOCK_TBL_SIZE; bx+=2) {
+        block_desc_t *blk = &(cfg->blk_tbl->tbl[bx]);
+        if(blk->local_addr == NULL) // ?
+            continue;
+
+        /* TODO Eventually we may have to unpin blocks, this will catch that.
+         * Right now this should never trigger. */
+        assert(blk->mr != NULL);
+
+        char* block = (char*)malloc(sizeof(char) * blk->size);
+        struct ibv_mr* block_mr = rmem_create_mr(block, sizeof(char) * blk->size);
+
+        int err = rmem_get(&(cfg->rmem), block, block_mr,
+                bx + 1, blk->size);
+        
+        RETURN_ERROR(err != 0, false,
+                ("Block %d does not match replica\n", bx));
+
+        err = memcmp(block, blk->local_addr, sizeof(char) * blk->size);
+        free(block);
+
+    }
+
+    free(block);
+    ibv_dereg_mr(block_mr);
+
+    return true;
+}
+
+//int rmem_get(struct rmem *rmem, void *dst, struct ibv_mr *dst_mr,
+//        uint64_t src, size_t size)
+
+bool rvm_txn_commit(rvm_cfg_t* cfg, rvm_txid_t txid)
 {
     /* TODO This is a brain-dead initial implementation. We simply copy over
      * everything in the block table, even if it wasn't modified. The long-term
      * solution is to use protection bits on the page table to automatically
      * detect changes and only copy those.
-     *
-     * TODO This also is not atomic. The current plan is to add a simple
-     * transaction to the rmem_put API.
```

Comment:
Yes, the last 3 lines of the comment (about the atomicity) were removed. The first 4 lines still hold and were not changed.